node-repository/src/main/java/com/yahoo/vespa/hosted/provision/maintenance/HostCapacityMaintainer.java

```diff
@@ -282,11 +272,42 @@ true, nodeRepository().nameResolver(), nodeRepository().nodes(), nodeRepository(
                                   nodeResources,
                                   nodeRepository().clock().instant()))
                 .toList();
-
     }
 
     private static NodeResources toNodeResources(ClusterCapacity clusterCapacity) {
-        return new NodeResources(clusterCapacity.vcpu(), clusterCapacity.memoryGb(), clusterCapacity.diskGb(),
-                clusterCapacity.bandwidthGbps());
+        return new NodeResources(clusterCapacity.vcpu(),
+                                 clusterCapacity.memoryGb(),
+                                 clusterCapacity.diskGb(),
+                                 clusterCapacity.bandwidthGbps(),
+                                 NodeResources.DiskSpeed.valueOf(clusterCapacity.diskSpeed()),
+                                 NodeResources.StorageType.valueOf(clusterCapacity.storageType()),
+                                 NodeResources.Architecture.valueOf(clusterCapacity.architecture()));
     }
+
+    private static List<Node> findEmptyOrRemovableHosts(List<Node> provisionedSnapshot) {
+        // Group nodes by parent; no parent means it's a host.
+        var nodesByParent = provisionedSnapshot.stream().collect(groupingBy(Node::parentHostname));
+
+        // Find all hosts that we once thought were empty (first clause), or whose children are now all removable (second clause).
+        return getHosts(nodesByParent).stream()
+                .filter(host -> host.hostEmptyAt().isPresent() || allChildrenCanBeDeprovisioned(nodesByParent, host))
+                .toList();
+    }
+
+    private static List<Node> getHosts(Map<Optional<String>, List<Node>> nodesByParent) {
+        return nodesByParent.get(Optional.<String>empty());
+    }
+
+    private static List<Node> getChildren(Map<Optional<String>, List<Node>> nodesByParent, Node host) {
+        return nodesByParent.getOrDefault(Optional.of(host.hostname()), List.of());
+    }
+
+    private static boolean allChildrenCanBeDeprovisioned(Map<Optional<String>, List<Node>> nodesByParent, Node host) {
+        return getChildren(nodesByParent, host).stream().allMatch(HostCapacityMaintainer::canDeprovision);
+    }
+
+    private static boolean someChildrenCannotBeDeprovisioned(Map<Optional<String>, List<Node>> nodesByParent, Node host) {
+        return  ! allChildrenCanBeDeprovisioned(nodesByParent, host);
```

Comment:
This will always return `false` if the given host has no children?

Reply:
Good catch, thanks, fixed